numpy/core/_objectmath.py

```diff
@@ -0,0 +1,262 @@
+"""
+Pure python implementations of ufuncs for Object type.
+Makes use of math module for some ufuncs.
+
+These are the last-resort fallback functions used to evaluate ufuncs on object
+array. Before these are used, numpy will first 1. look for a Python built-in
+function corresponding to the ufunc, and call it 2. try to convert built-in
+python scalars to numpy types, then use of the ufunc, and 3.  check for a
+user-defined method on the item with the same name as the ufunc.
+
+"""
+from __future__ import division, absolute_import, print_function
+import math
+from numpy.core.umath import pi as PI
+import numpy as np
+import sys
+
+LOG2 = math.log(2)
+
+if sys.version_info > (3,):
+    def isint(x):
+        return type(x) in (bool, int)
+else:
+    def isint(x):
+        return type(x) in (bool, int, long)
+
+def float_or_complex(x):
+    return type(x) in (float, complex)
+
+def _notimplemented_msg(fname, etype, etype2=None):
+    if etype2 is not None:
+        typestr = "({a}, {b})".format(a=etype, b=etype2)
+    else:
+        typestr = str(etype)
+    return ("object-array fallback for ufunc '{name}' is not implemented for "
+            "arguments of type {type}").format(name=fname, type=typestr)
+
+
+def logical_and(x, y):
+    return bool(x and y)
+
+def logical_or(x, y):
+    return bool(x or y)
```

Comment:
I'd be much keener to leave behaviour changes to a second PR, if possible - in particular, we might want a deprecation cycle, or even a way to explicitly request the old behaviour. But I guess we can review it all together, and split it up once everything else looks good.